sorc/ncep_post.fd/UPP_PHYSICS.f

```diff
@@ -4498,5 +4501,454 @@ END SUBROUTINE CALSLR_UUTAH
+     &                      -  (VWND(I,jj-1)*COSL(I,Jj-1)                       &
+     &                      -   VWND(I,J)*COSL(I,J))*wrk3(i,jj)) * wrk1(i,jj) 
+              enddo
+            ENDIF
+          endif
+        ELSE
+          DO I=ISTA,IEND
+            ip1 = ie(i)
+            im1 = iw(i)
+            if(VWND(ip1,J)==SPVAL .or. VWND(im1,J)==SPVAL .or. &
+               UWND(I,J-1)==SPVAL .or. UWND(I,J+1)==SPVAL) cycle
+            ABSV(I,J)   = ((VWND(ip1,J)-VWND(im1,J))*wrk2(i,j)                 &
+     &                    -  (UWND(I,J-1)*COSL(I,J-1)                          &
+                          -   UWND(I,J+1)*COSL(I,J+1))*wrk3(i,j)) * wrk1(i,j)  &
+                          + F(I,J)
+            DIV(I,J)    = ((UWND(ip1,J)-UWND(im1,J))*wrk2(i,j)                 &
+     &                    +  (VWND(I,J-1)*COSL(I,J-1)                          &
+                          -   VWND(I,J+1)*COSL(I,J+1))*wrk3(i,j)) * wrk1(i,j)  
+          ENDDO
+        END IF
+        if (npass > 0) then
+          do i=ista,iend
+            tx1(i) = absv(i,j)
+          enddo
+          do nn=1,npass
+            do i=ista,iend
+              tx2(i+1) = tx1(i)
+            enddo
+            tx2(1)    = tx2(im+1)
+            tx2(im+2) = tx2(2)
+            do i=2,im+1
+              tx1(i-1) = 0.25 * (tx2(i-1) + tx2(i+1)) + 0.5*tx2(i)
+            enddo
+          enddo
+          do i=ista,iend
+            absv(i,j) = tx1(i)
+          enddo
+        endif
+      END DO                               ! end of J loop
 
+!     deallocate (wrk1, wrk2, wrk3, cosl)
+! GFS use lon avg as one scaler value for pole point
+!
+      ! call poleavg(IM,JM,JSTA,JEND,SMALL,COSL(1,jsta),SPVAL,ABSV(1,jsta))
+!
+      call exch(absv(ista_2l:iend_2u,jsta_2l:jend_2u))
+      call fullpole(absv(ista_2l:iend_2u,jsta_2l:jend_2u),avpoles)     
+!
+      cosltemp=spval
+      if(jsta== 1) cosltemp(1:im, 1)=coslpoles(1:im,1)
+      if(jend==jm) cosltemp(1:im,jm)=coslpoles(1:im,2)
+      avtemp=spval
+      if(jsta== 1) avtemp(1:im, 1)=avpoles(1:im,1)
+      if(jend==jm) avtemp(1:im,jm)=avpoles(1:im,2)
+!        
+      call poleavg(IM,JM,JSTA,JEND,SMALL,cosltemp(1,jsta),SPVAL,avtemp(1,jsta))
+!
+      if(jsta== 1) absv(ista:iend, 1)=avtemp(ista:iend, 1)
+      if(jend==jm) absv(ista:iend,jm)=avtemp(ista:iend,jm)
+!    
+!        deallocate (wrk1, wrk11, wrk2, wrk3, cosl, iw, ie)
+!    
+      call exch(absv(ista_2l:iend_2u,jsta_2l:jend_2u))
+      call exch(div(ista_2l:iend_2u,jsta_2l:jend_2u))
+!
+! store absv and div factors before poisson loops
+!$omp parallel do private(i,j)
+      DO J=JSTA,JEND
+      DO I=ISTA,IEND
+        ATMP(I,J)=0.25*(ABSV(I,J)-F(I,J))/(wrk2(i,j)*wrk1(i,j)*wrk3(i,j)*wrk1(i,j)*COSL(i,j)*4.)
+        DTMP(I,J)=0.25*DIV(I,J)/(wrk2(i,j)*wrk1(i,j)*wrk3(i,j)*wrk1(i,j)*COSL(i,j)*4.)
+      ENDDO
+      ENDDO
+!
+! poisson solver for psi and chi 
+      PSI=0.
+!$omp parallel do private(jjk,jj)
+      do jjk=1,500
```

Comment:
@WenMeng-NOAA Using only one loop, the results are the same but the runtimes improve slightly (from 5 to 6 sec for SFS and from 2:32 to 2:44 min for GFS). It may have something to do with starting each jjk loop at a better value than 0. Would you prefer keeping the 2 loops or use only 1?

Reply:
@KarinaAsmar-NOAA 
Please remove the following lines of omp in both PSI and CHI blocks,
!$omp parallel do private(jjk,jj)
These loops are iterations that parallelization does not apply.